File: bifrost-ts/src/auth/server/cookie-utils.ts
import type { CookieOptions } from '../types'

/**
 * Response-like object that can set headers.
 */
interface ResponseLike {
  setHeader: (name: string, value: string) => void
}

/**
 * Request-like object that may contain cookies.
 */
interface RequestLike {
  cookies?: Record<string, string>
  headers?: { cookie?: string }
}

/**
 * Set an HttpOnly refresh token cookie.
 *
 * @param res - Response object with setHeader method
 * @param token - Refresh token value
 * @param options - Cookie configuration
 */
export function setRefreshTokenCookie(
  res: ResponseLike,
  token: string,
  options: CookieOptions,
): void {
  const maxAge = options.maxAgeDays * 24 * 60 * 60
  const secure = options.secure ?? process.env.NODE_ENV === 'production'
  /**
   * Lax prevents CSRF on cross-site POST while still sending the cookie
   * on same-site navigations and Vite HMR reloads. Strict caused the
   * refresh cookie to be dropped during dev hot reloads.
   */
  const sameSite = options.sameSite ?? 'Lax'
  const path = options.path ?? '/'

  // URL-encode token to handle special characters (;, =, etc.)
  const encodedToken = encodeURIComponent(token)

  const parts = [
    `${options.name}=${encodedToken}`,
    'HttpOnly',
    `Path=${path}`,
    `Max-Age=${maxAge}`,
    `SameSite=${sameSite}`,
  ]

  if (secure) {
    parts.push('Secure')
  }

  res.setHeader('Set-Cookie', parts.join('; '))
}

/**
 * Clear a refresh token cookie by setting it to expire immediately.
 *
 * @param res - Response object with setHeader method
 * @param options - Cookie configuration (name, secure, path)
 */
export function clearRefreshTokenCookie(
  res: ResponseLike,
  options: Pick<CookieOptions, 'name' | 'secure' | 'path'>,
): void {
  const secure = options.secure ?? process.env.NODE_ENV === 'production'
  const path = options.path ?? '/'

  const parts = [
    `${options.name}=`,
    'HttpOnly',
    `Path=${path}`,
    'Max-Age=0',
    'SameSite=Lax',
  ]

  if (secure) {
    parts.push('Secure')
  }

  res.setHeader('Set-Cookie', parts.join('; '))
}

/**
 * Extract refresh token from a request.
 * Checks parsed cookies first, then falls back to manual cookie header parsing.
 * Automatically URL-decodes the cookie value.
 *
 * @param req - Request object with optional cookies or cookie header
 * @param cookieName - Name of the cookie to extract
 * @returns Cookie value if found, undefined otherwise
 */
export function getRefreshTokenFromRequest(
  req: RequestLike,
  cookieName: string,
): string | undefined {
  // Check parsed cookies first (if cookie-parser middleware was used)
  // cookie-parser auto-decodes, so no need to decode again
  if (req.cookies?.[cookieName]) {
    return req.cookies[cookieName]
  }

  // Fall back to manual parsing from cookie header
  const cookieHeader = req.headers?.cookie
  if (!cookieHeader) {
    return undefined
  }

  // Escape regex special characters in cookie name to prevent injection
  const escapedName = cookieName.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
  const match = cookieHeader.match(
    // eslint-disable-next-line security/detect-non-literal-regexp -- pattern is escaped above
    new RegExp(`(?:^|;\\s*)${escapedName}=([^;]*)`),
  )

  if (!match) {
    return undefined
  }

  // URL-decode the cookie value (matches our encoding in setRefreshTokenCookie)
  try {
    return decodeURIComponent(match[1])
  } catch {
    // If decode fails, return raw value (backwards compatibility)
    return match[1]
  }
}

File: typeferry-py/src/typeferry/auth/cookie_utils.py
"""Cookie helpers — port of ``typeferry-ts/src/auth/server/cookie-utils.ts``.

The Set-Cookie serialization MUST match the TS output byte-for-byte so
the JS client's refresh flow works transparently with a Python server
(PROTOCOL.md §8.3).
"""

from __future__ import annotations

import os
import re
from urllib.parse import quote, unquote

from typeferry.auth.types import CookieOptions


def _default_secure() -> bool:
    return os.environ.get("NODE_ENV") == "production"


def set_refresh_token_cookie(token: str, options: CookieOptions) -> str:
    """Return the ``Set-Cookie`` header value encoding ``token``.

    Transport adapters call this helper and attach the returned string
    as a ``Set-Cookie`` response header.
    """

    max_age = options.max_age_days * 24 * 60 * 60
    secure = options.secure if options.secure is not None else _default_secure()
    same_site = options.same_site
    path = options.path

    encoded = quote(token, safe="")
    parts = [
        f"{options.name}={encoded}",
        "HttpOnly",
        f"Path={path}",
        f"Max-Age={max_age}",
        f"SameSite={same_site}",
    ]
    if secure:
        parts.append("Secure")
    return "; ".join(parts)


def clear_refresh_token_cookie(
    name: str, *, secure: bool | None = None, path: str = "/"
) -> str:
    """Return a ``Set-Cookie`` header that clears the named cookie."""

    resolved_secure = secure if secure is not None else _default_secure()
    parts = [
        f"{name}=",
        "HttpOnly",
        f"Path={path}",
        "Max-Age=0",
        "SameSite=Lax",
    ]
    if resolved_secure:
        parts.append("Secure")
    return "; ".join(parts)


def get_refresh_token_from_cookie_header(
    cookie_header: str | None, cookie_name: str
) -> str | None:
    """Extract a cookie value from a raw ``Cookie:`` header.

    Returns None when the cookie is absent. URL-decodes the value to
    mirror :func:`set_refresh_token_cookie` encoding.
    """

    if not cookie_header:
        return None
    escaped = re.escape(cookie_name)
    match = re.search(rf"(?:^|;\s*){escaped}=([^;]*)", cookie_header)
    if match is None:
        return None
    value = match.group(1)
    try:
        return unquote(value)
    except Exception:
        return value
